_topics/en/cluster-mempool.md

```diff
@@ -52,20 +52,98 @@ see_also:
+set formed by a single transaction and all its unconfirmed ancestors).
+Eviction removes transactions in order of lowest-descendant feerate
+first (feerate of a set formed by a single unconfirmed transaction and
+its descendants). This mechanism is certainly suboptimal (highest
+ancestor-feerate first is just an approximation for highest-feerate
+subset overall, and e.g. can't deal with
+multiple-children-pay-for-the-same-parent), but more problematic is the
+fact that eviction isn't the exact opposite of mining: they're both
+approximations, and the ways they are suboptimal don't cancel out, so
+they don't result in the opposite order of each other.
+
+## The obvious solution doesn't work
+
+There is an obvious solution to the problem: instead of finding
+lowest-feerate-descendant-sets, run the normal mining algorithm on the
+entire mempool (don't stop after one block worth of transactions), and
+see what it would include last. That's the thing you should evict!
+
+Sadly, this is computationally infeasible. The block building algorithm
+is fairly fast when running on one block worth of transactions, but
+running it on the entire mempool would take a significant multiple of
+the time. This is absolutely not possible to do every time any time
+something needs to be evicted (which may be multiple times per second,
+and ideally, with ~millisecond latency to not stall other processing).
+
+An obvious question arises: can't we precompute things so that computing
+the updated mining-order-of-full-mempool isn't too slow? And we kind of
+can. What the block building algorithm ultimately does is find groups of
+transactions to include at once (e.g. child pays for parent means both
+will be included at once), at the feerate that included set overall has
+(sum of fees divided by sum of sizes). If we could somehow precompute
+those groupings (which we'll call chunks) ahead of time, then the actual
+block building algorithm (ignoring binpacking issues once we get close
+to the block being full) is just including chunks in decreasing feerate
+order.
+
+## Precomputing works if the only thing we chunk are clusters
+
+So can we precompute the chunks, and update just the ones that get
+affected when a new transaction or block comes in? Sadly, there is in
+theory no bound on how many transactions' chunkings can get affected by
+even a single new transaction---it could be the entire mempool in which
+case we'd be back to recomputing everything. See [this example][wuille
+slide11] where a single new added transaction completely reverses the
+chunking of all transactions.
+
+However, it is hopefully apparent that the limit of how much can be
+affected by a transaction is just whatever transactions it is directly
+or indirectly connected to, i.e. its cluster.<!--FIXME:harding confirm edit--> Optimal chunks never cross
+cluster boundaries---any chunk that did could be split into independent
+chunks, and doing so would never worsen the result.
```

Comment:
Updated the intro paragraph to the text below:

>   **Cluster mempool** is a proposal to associate each unconfirmed
  transaction in a mempool with related transactions, creating a
  _cluster_.  Each cluster contains feerate-sorted _chunks_ consisting
  of one or more transactions.  Thus, if clusters are limited in size,
  we also limit how much needs to be recomputed in response to new
  transactions being added to the mempool, allowing algorithmic
  decisions affecting the entire mempool to to complete fast enough to
  use them in P2P network code.

(The other revisions were required because we limit the first-paragraph to 500 characters in order to be able to use them in the future in something like a glossary.)